docs/architecture/adr-0025-user-access-scopes.md

```diff
@@ -0,0 +1,147 @@
+
+- It is assumed that a majority of users will not have any access restrictions
+- In the most common application of this system (preventing project creation), specifying what _can_ be done would require _many_ more rules than specifying what _can't_ be done
+
+### The Basic Solution
+
+- Each user has a column representing a list of read-only scopes for that user (i.e. for some user the column named `scopes` has a list containing `PROJECT` and `SCENE`)
+- Every value in the scope list is a member of `ObjectType`
+- Each API endpoint that is affected would do a lookup on the requesting user and see if there was a scope present that should prevent the request from being fulfilled.
+
+Usage at the endpoint level might look like:
+
+```scala
+  def createProject: Route = authenticate { user =>
+    authorizeAsync {
+      UserDao
+        .withoutScoping(ObjectType.Project)
+        .transact(xa)
+        .unsafeToFuture
+    } {
+      entity(as[Project.Create]) { newProject =>
+        onSuccess(
+          ProjectDao
+            .insertProject(newProject, user)
+            .transact(xa)
+            .unsafeToFuture) { project =>
+          complete(StatusCodes.Created, project)
+        }
+      }
+    }
+  }
+```
+
+#### Advantages
+
+- no possible conflicts as this only handles read-only restrictions
+- minimal effect on current data-models -- requires one additional field on `User`
+
+#### Disadvantages
+
+- only supports read-only scoping
+- requires some machinery to encode/decode scopes to/from the DB
+
+### The Wild West Solution
+
+This solution has almost no constraints.
+
+- Each user has a column representing a list of scopes, for example `ProjectReadOnly` or `ProjectNoDelete`
+- These scopes could be defined within a `ScopeType` to enforce values
+- Each API endpoint that is affected would do a lookup on the requesting user and see if there was a scope present that should prevent the request from being fulfilled.
+
+Usage at the endpoint level might look like:
+
+```scala
+  def createProject: Route = authenticate { user =>
+    authorizeAsync {
+      UserDao
+        .withoutScoping(ScopeType.ProjectReadOnly)
+        .transact(xa)
+        .unsafeToFuture
+    } {
+      entity(as[Project.Create]) { newProject =>
+        onSuccess(
+          ProjectDao
+            .insertProject(newProject, user)
+            .transact(xa)
+            .unsafeToFuture) { project =>
+          complete(StatusCodes.Created, project)
+        }
+      }
+    }
+  }
+```
+
+#### Advantages
+
+- easy to add new scopes as needed
+- minimal effect on current data-models -- requires one additional field on `User`
+
+#### Disadvantages
+
+- lack of constraints make it possible to have conflicting rules which would necessitate some programmatic resolution approach
```

Comment:
What sorts of constraints do you imagine here? I'm not sure I grasp where they come from from the description

Reply:
That's sort of the point? There aren't any constraints which allows there to be conflicts. I feel like I might be missing something with your comment though.

Reply:
Based on what we're discussing in meatspace, this stems from different understandings of what a scope should be. If we understand a scope as `project:readOnly`, then that scope requires us to understand that we shouldn't later be able to add `project:create` without first eliminating `project:readOnly`, or our scopes would be in conflict. However if we instead understand no project scopes as `project:*` and any other scopes as an explicit list of the things you're allowed to do with projects, we I think are just as expressive without any risk of conflicts.

UX here can get a bit annoying -- and it's possible to end up in surprising circumstances once the permissions model gets pretty complex (e.g. #4825), but I think we can avoid conflicts entirely as long as scopes don't contain information about other allowed scopes.

Reply:
I'm going to update the ADR with an implementation that looks like this suggestion.